### DISCUSSION TITLE
Gurobi: non-linear monomial

### DISCUSSION THREAD

**Discussion by zoharl3** [yalmip/YALMIP#1076] [General discussion] [answered]
In the code below, what does yalmip do with the first constraint? Mosek, of course, doesn't accept this.

```
x = intvar(5, 1)
b = binvar(5, 1);
z = intvar(1);

con = [
    b'*x <= z;
    x >= 3*b;
    sum(b) == 3;
];

options = sdpsettings('solver', 'gurobi');

obj = z

sol = optimize( con, obj, options );
```

**Comment by johanlofberg** [comment, marked as answer]
mosek does not supprt this bilinear inequality (although it is easily linearizable), gurobi does.

**Reply by zoharl3** [reply]
So gurobi accepts bilinear ineq.

Just to elaborate on it's "easily linearizable" for future reference.

If binary vars are involved (i.e. MILP), then I can believe it--you can do almost anything--something in the spirit of:

http://yetanothermathprogrammingconsultant.blogspot.com/2008/05/multiplication-of-continuous-and-binary.html

or 

https://math.stackexchange.com/questions/838340/linearization-of-a-product-of-two-decision-variables

which refers to

https://download.aimms.com/aimms/download/manuals/AIMMS3OM_IntegerProgrammingTricks.pdf

---

I noticed, however, that even if I define the vars as continuous (sdpvar), then gurobi is happy. However, it says not convex, using MI. So, it probably adds binvars  and solves as above.

**Reply by johanlofberg** [reply]
Yes
https://yalmip.github.io/tutorial/logicprogramming#products

Gurobi solves general nonconvex quadratically constrained quadratic programs

**Reply by zoharl3** [reply]
Nice cheat sheet. I've been figuring it out and doing it "manually" all this time...
